File: CsharpProjects/StateSolucao01/StateSolucao01/domains/Pedido.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;
using System.Xml.Serialization;

using StateSolucao01.domains.enums;

namespace StateSolucao01.domains
{
    public class Pedido
    {
        private StatusPedido statusPedido { get; set; }

        public Pedido() {
            Console.WriteLine("Pedido Aguardando Pagamento");
            this.statusPedido = StatusPedido.AGUARDANDO_PAGAMENTO;
        }

        public void sucessoAoPagar()
        {
            if (this.statusPedido == StatusPedido.AGUARDANDO_PAGAMENTO)
            {
                //muda status
                Console.WriteLine("Pedido Pago");
                this.statusPedido = StatusPedido.PAGO;
            }
            else
            {
                throw new Exception("O pedido não está aguardando pagamento");
            }
        }

        public void cancelarPedido()
        {
            if (this.statusPedido == StatusPedido.PAGO ||
                this.statusPedido == StatusPedido.AGUARDANDO_PAGAMENTO)
            {
                //muda status
                Console.WriteLine("Pedido Cancelado");
                this.statusPedido = StatusPedido.CANCELADO;
            }
            else
            {
                throw new Exception("O pedido se encontra "+this.statusPedido.ToString());
            }
        }

        public void despacharPedido()
        {
            if(this.statusPedido == StatusPedido.PAGO)
            {
                //muda status
                Console.WriteLine("Pedido Enviado");
                this.statusPedido = StatusPedido.ENVIADO;
            } else
            {
                throw new Exception("O pedido se encontra cancelado");
            }
        }

    }
}

File: JavaProjects/StateSolucao01/src/domains/Pedido.java
/*
 * Click nbfs://nbhost/SystemFileSystem/Templates/Licenses/license-default.txt to change this license
 * Click nbfs://nbhost/SystemFileSystem/Templates/Classes/Class.java to edit this template
 */
package domains;

import domains.enums.StatusPedido;
/**
 *
 * @author jeffe
 */
public class Pedido {
    
    private StatusPedido statusPedido;

    public Pedido() {
        this.statusPedido = StatusPedido.AGUARDANDO_PAGAMENTO;
    }
    
    public void sucessoAoPagar() throws Exception
    {
        if (this.statusPedido == StatusPedido.AGUARDANDO_PAGAMENTO)
        {
            //muda status
            System.out.println("Pedido Pago");
            this.statusPedido = StatusPedido.PAGO;
        }
        else
        {
            throw new Exception("O pedido não está aguardando pagamento");
        }
    }

    public void cancelarPedido() throws Exception
    {
        if (this.statusPedido == StatusPedido.PAGO ||
            this.statusPedido == StatusPedido.AGUARDANDO_PAGAMENTO)
        {
            //muda status
            System.out.println("Pedido Cancelado");
            this.statusPedido = StatusPedido.CANCELADO;
        }
        else
        {
            throw new Exception("O pedido se encontra "+this.statusPedido.toString());
        }
    }

    public void despacharPedido() throws Exception
    {
        if(this.statusPedido == StatusPedido.PAGO)
        {
            //muda status
            System.out.println("Pedido Enviado");
            this.statusPedido = StatusPedido.ENVIADO;
        } else
        {
            throw new Exception("O pedido se encontra cancelado");
        }
    }
}
